src/systemctl/systemctl-list-units.c

```diff
@@ -105,11 +107,205 @@ static char *format_unit_id(const char *unit, const char *machine) {
+        LIST_FIELDS(struct MapNode, hashmap);
+} MapNode;
+
+/* Hash map definition */
+MapNode *hash_map[MAP_SIZE] = {NULL};
+
+/* Function to calculate hash value of the slice name */
+static size_t hash(const char *slice_name) {
+        assert(slice_name);
+
+        size_t hash_value = 0;
+        for (const char *p = slice_name; *p != '\0'; ++p)
+                hash_value = 31 * hash_value + (*p);
+
+        return hash_value % MAP_SIZE;
+}
+
+/* Function to insert or update units under slice unit in the hash map */
+static int insert_units_under_slice(const char *slice_name, char **units, size_t num_units) {
+        size_t index;
+        MapNode *new_node, *current = NULL;
+
+        assert(slice_name);
+        assert(units);
+        assert(num_units);
+        index = hash(slice_name);
+
+        /* Create a new hash map node */
+        new_node = new(MapNode, sizeof(MapNode));
+        if (new_node == NULL)
+                return log_oom();
+
+        /* Copy slice name and units */
+        strcpy(new_node->slice_name, slice_name);
+        /* Allocate memory for the units array and copy content into it */
+        new_node->units_under_slice.units = new(char *, num_units);
+        if (new_node->units_under_slice.units == NULL) {
+                mfree(new_node);
+                return log_oom();
+        }
+        for(size_t i = 0; i < num_units; i++) {
+                new_node->units_under_slice.units[i] = strdup(units[i]);
+                if (new_node->units_under_slice.units[i] == NULL) {
+                        for (size_t j = 0; j < i; j++)
+                                free(new_node->units_under_slice.units[j]);
+
+                        mfree(new_node->units_under_slice.units);
+                        mfree(new_node);
+                        return log_oom();
+                }
+
+        }
+        new_node->units_under_slice.num_units = num_units;
+        new_node->hashmap_next = NULL;
+
+        /* Insert or update the node in the hash map */
+        if (hash_map[index] == NULL)
+                hash_map[index] = new_node;
+        else {
+                /* Append new node to end of the list */
+                current = hash_map[index];
+                while(current->hashmap_next != NULL)
+                        current = current->hashmap_next;
+                current->hashmap_next = new_node;
+        }
+        return 0;
+
+}
+
+/* Function to free memory allocation for units under a slice */
+static void free_units_under_slice(UnitsUnderSlice *units_under_slice) {
+        for (size_t i = 0; i < units_under_slice->num_units; i++) {
+                mfree(units_under_slice->units[i]);
+        }
+        mfree(units_under_slice->units);
+}
+
+static void free_hash_map(void) {
+        MapNode *current = NULL, *next = NULL;
+
+        for(size_t i = 0; i < MAP_SIZE; i++) {
+              current =  hash_map[i];
+              while (current != NULL) {
+                next = current->hashmap_next;
```

Comment:
indentation and formatting are wrong, look at docs/CODING_STYLE.md and adjust here and elsewhere